### PR TITLE
Deterministically dispose instances created by WebHostBuilder

### PR DESCRIPTION
**Pull request by pakrym**
Fixes: https://github.com/aspnet/Hosting/issues/867
IServiceProvider would stop disposing instances: https://github.com/aspnet/DependencyInjection/pull/462

@davidfowl @Tratcher

### CODE REVIEW

File: src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs [hunk 1]
```diff
@@ -195,9 +195,8 @@ private IServiceCollection BuildHostingServices()
             services.AddTransient<IHttpContextFactory, HttpContextFactory>();
             services.AddOptions();
 
-            var diagnosticSource = new DiagnosticListener("Microsoft.AspNetCore");
-            services.AddSingleton<DiagnosticSource>(diagnosticSource);
-            services.AddSingleton<DiagnosticListener>(diagnosticSource);
+            services.AddSingleton<DiagnosticListener>(provider => new DiagnosticListener("Microsoft.AspNetCore"));
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs]: Does this need to be disposed?

File: src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs [hunk 2]
```diff
@@ -257,5 +256,10 @@ private string ResolveContentRootPath(string contentRootPath, string basePath)
             }
             return Path.Combine(Path.GetFullPath(basePath), contentRootPath);
         }
+
+        public void Dispose()
+        {
+            _loggerFactory?.Dispose();
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs]: Shouldn't this be disposed with the host? Why is the builder disposable?

File: src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs [hunk 3]
```diff
@@ -195,9 +195,8 @@ private IServiceCollection BuildHostingServices()
             services.AddTransient<IHttpContextFactory, HttpContextFactory>();
             services.AddOptions();
 
-            var diagnosticSource = new DiagnosticListener("Microsoft.AspNetCore");
-            services.AddSingleton<DiagnosticSource>(diagnosticSource);
-            services.AddSingleton<DiagnosticListener>(diagnosticSource);
+            services.AddSingleton<DiagnosticListener>(provider => new DiagnosticListener("Microsoft.AspNetCore"));
+            services.AddSingleton<DiagnosticSource>(provider => provider.GetService<DiagnosticListener>());
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs]: Not sure I like this.

File: src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs (hunk 1 above)
Comment (pakrym) [inline src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs]: It is IDisposible

File: src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs (hunk 2 above)
Comment (pakrym) [inline src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs]: No, Builder creates it and will reuse for any number of WebHosts you build using it.

File: src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs (hunk 3 above)
Comment (pakrym) [inline src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs]: We don't have multi interface registration and there is no need to have this as an instance as far as I understand.

File: src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs [hunk 4]
```diff
@@ -22,7 +22,7 @@ namespace Microsoft.AspNetCore.Hosting
     /// <summary>
     /// A builder for <see cref="IWebHost"/>
     /// </summary>
-    public class WebHostBuilder : IWebHostBuilder
+    public class WebHostBuilder : IWebHostBuilder, IDisposable
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs]: Should IWebHostBuilder be IDisposable?

**Comment by pakrym** [comment]
1. Changed WebHostBuilder to allow creating only single instance of WebHost.
2. LoggerFactory is now disposed with WebHost is it was automatically created.
3. If LoggerFactory was passed in it's not disposed.
4. Added tests for all these cases.

File: src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs (hunk 1 above)
Comment (pakrym) [inline src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs]: then we would need two different `BuildHostingServices` or at least parts that need to chain.

Comment (pakrym) [inline src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs]: Also we don't mutate `DiagnosticSource` so it might be fine to return different instances.

**Comment by pakrym** [comment]
Updated again, now there is two methods to add host services and to forward them to application services. Also fixed server disposal issue.

File: src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs [hunk 5]
```diff
@@ -245,6 +262,17 @@ private IServiceCollection BuildHostingServices()
...
+        private void AddApplicationServices(IServiceCollection services, IServiceProvider hostingServiceProvider)
+        {
+            var loggerFactory = hostingServiceProvider.GetService<ILoggerFactory>();
+            services.Replace(ServiceDescriptor.Singleton(typeof(ILoggerFactory), loggerFactory));
+
+            var listener = hostingServiceProvider.GetService<DiagnosticListener>();
+            services.Replace(ServiceDescriptor.Singleton(typeof(DiagnosticListener), listener));
+            services.Replace(ServiceDescriptor.Singleton(typeof(DiagnosticSource), listener));
+        }
+
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs]: nit: extra space

File: src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs [hunk 6]
```diff
@@ -245,6 +262,17 @@ private IServiceCollection BuildHostingServices()
             return services;
         }
 
+        private void AddApplicationServices(IServiceCollection services, IServiceProvider hostingServiceProvider)
+        {
+            var loggerFactory = hostingServiceProvider.GetService<ILoggerFactory>();
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs]: Add comments here explaining why thungs are so bizzare

File: src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs [hunk 7]
```diff
@@ -175,30 +189,33 @@ private IServiceCollection BuildHostingServices()
             var services = new ServiceCollection();
             services.AddSingleton(_hostingEnvironment);
 
+            //The configured ILoggerFactory is added as a singleton here. AddLogging below will not add an additional one.
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs]: nit: space after //

**Comment by pakrym** [comment]
Also undid change to `UseServer` and fixed tests for it.

File: test/Microsoft.AspNetCore.Hosting.Tests/WebHostBuilderTests.cs [hunk 8]
```diff
@@ -188,6 +189,7 @@ public void MultipleConfigureLoggingInvokedInOrder()
                 .UseStartup<StartupNoServices>();
 
             var host = (WebHost)hostBuilder.Build();
+            var factory = host.Services.GetService<ILoggerFactory>();
```
Comment (davidfowl) [inline test/Microsoft.AspNetCore.Hosting.Tests/WebHostBuilderTests.cs]: What is this for?

File: test/Microsoft.AspNetCore.Hosting.Tests/WebHostBuilderTests.cs [hunk 9]
```diff
@@ -494,6 +496,86 @@ public void Configure_SupportsStaticMethodDelegate()
...
+
+            var host = builder.UseServer(server)
+                .ConfigureServices(collection => collection.AddSingleton<ILoggerFactory>(factory))
+                .UseStartup<StartupWithILoggerFactory>()
+                .Build();
+
+            var factoryFromHost = host.Services.GetService<ILoggerFactory>();
+            Assert.Equal(factory, factoryFromHost);
+        }
+
```
Comment (davidfowl) [inline test/Microsoft.AspNetCore.Hosting.Tests/WebHostBuilderTests.cs]: nit: extra spaces

Comment (pakrym) [inline test/Microsoft.AspNetCore.Hosting.Tests/WebHostBuilderTests.cs]: Once `ILoggerFactory` was registered as factory so resolve was required to invoke callbacks. Not true anymore. Reverting.

File: src/Microsoft.AspNetCore.Hosting/Internal/ServiceCollectionExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,20 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Hosting.Internal
+{
+    public static class ServiceCollectionExtensions
```
Comment (muratg) [inline src/Microsoft.AspNetCore.Hosting/Internal/ServiceCollectionExtensions.cs]: Actually making the class internal?

**Pull request merged**